group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -3575,6 +3600,65 @@ public void replay(
         }
     }
 
+    /**
+     * Replays ConsumerGroupRegexKey/Value to update the hard state of
+     * the consumer group related to a regular expression. It updates
+     * the resolution of the regular expression with the value, or deletes
+     * the regular expression if the record value is null.
+     *
+     * @param key   A ConsumerGroupRegexKey record.
+     * @param value A ConsumerGroupRegexValue record.
+     */
+    public void replay(
+        ConsumerGroupRegexKey key,
+        ConsumerGroupRegexValue value
+    ) {
+
+        String groupId = key.groupId();
+        String regex = key.regex();
+
+        try {
+            getOrMaybeCreatePersistedConsumerGroup(groupId, value != null);
+        } catch (GroupIdNotFoundException ex) {
+            // If the group does not exist anymore do not load regex.
+            // TODO: validate if this could be a case: group does exist anymore but regex record being replayed.
+            return;
+        }
+
+        if (value != null) {
+            updateGroupRegex(
+                new ConsumerGroupRegex.RegexKey.Builder().updateWith(key).build(),
+                new Resolution.Builder().updateWith(value).build()
+            );
```

Comment:
I wonder if we should rather store the resolved regex in the ConsumerGroup object in order to keep all the metadata about a consumer group as the same place. It is easy to forget about it otherwise. What do you think?

Reply:
yes, makes total sense, done and it does look much cleaner. The regexManger only cares about the async operation (resolves, writes records), info stored in the group (replayed from records). Makes sense?